docs/protocols/xlang_object_graph_spec.md

```diff
@@ -0,0 +1,747 @@
+
+For languages which doesn't support reference such as rust, reference tracking must be disabled for correct
+deserialization by fury rust implementation.
+
+For languages whose object values are not null by default:
+
+- In rust, Fury takes `Option:None` as a null value
+- In c++, Fury takes `std::nullopt` as a null value
+- In golang, Fury takes `null interface/pointer` as a null value
+
+If one want to deserialize in languages like `Java/Python/JavaScript`, he should mark the type with all fields
+not-null by default, or using schema-evolution mode to carry the not-null fields info in the data.
+
+## Type Meta
+
+For every type to be serialized, it must be registered with an optional ID first. The registered type will have a
+user-provided or an auto-growing unsigned int i.e. `type_id`. The registration can be used for security check and type
+identification. The id of user registered type will be added by `64` to make space for Fury internal data types.
+
+Depending on whether meta share mode and registration is enabled for current type, Fury will write type meta
+differently.
+
+### Schema consistent
+
+If schema consistent mode is enabled globally or enabled for current type, type meta will be written as a fury unsigned
+varint of `type_id`.
+
+### Schema evolution
+
+If schema evolution mode is enabled globally or enabled for current type, type meta will be written as follows:
+
+- Normal mode(meta share not enabled):
+    - If type meta hasn't been written before, add `type def`
+      to `captured_type_defs`: `captured_type_defs[type def] = map size`
+    - Get index of the meta in `captured_type_defs`, write that index as an unsigned varint.
+    - The meta will be written as:
+      ```
+      | unsigned varint: written index |
+      ```
+    - After finished the serialization of the object graph, fury will start to write `captured_type_defs`:
+        - Firstly, set current to `meta start offset` of fury header
+        - Then write `captured_type_defs` one by one:
+          ```python
+          buffer.write_var_uint32(len(writting_type_defs))
+          for type_meta in writting_type_defs:
+          type_meta.write_type_def(buffer)
+          ```
+- Meta share mode: the writing steps are same as the normal mode, but `captured_type_defs` will be shared across
+  multiple serializations of different objects. For example, suppose we have a batch to serialize:
+    ```python
+    captured_type_defs = {}
+    stream = ...
+    # add `Type1` to `captured_type_defs` and write `Type1`
+    fury.serialize(stream, [Type1()])
+    # add `Type2` to `captured_type_defs` and write `Type2`, `Type1` is written before.
+    fury.serialize(stream, [Type1(), Type2()])
+    # `Type1` and `Type2` are written before, no need to write meta.
+    fury.serialize(stream, [Type1(), Type2()])
+    ```
+
+- Streaming mode(streaming mode doesn't support meta share):
```

Comment:
How can we detect which mode is used? I noticed that there is no flag in the header to indicate the mode

Reply:
This is configured globally when creating fury. All the fury instance across multiple languages must have same config